framework/src/node/consensus/certificate_generation/commit_pool.ts

```diff
@@ -62,15 +64,103 @@ export class CommitPool {
 
 	// eslint-disable-next-line @typescript-eslint/no-empty-function
 	public async job(): Promise<void> {}
-	// eslint-disable-next-line @typescript-eslint/no-empty-function
-	public addCommit(_commit: SingleCommit, _height: number): void {}
-	// eslint-disable-next-line @typescript-eslint/no-empty-function
-	public validateCommit(_commit: SingleCommit): boolean {
+
+	public addCommit(commit: SingleCommit, height: number): void {
+		const currentCommits = this._nonGossipedCommits.get(height) ?? [];
+		const doesCommitExist = currentCommits.some(aCommit =>
+			aCommit.certificateSignature.equals(commit.certificateSignature),
+		);
+		if (!doesCommitExist) {
+			this._nonGossipedCommits.set(height, [...currentCommits, commit]);
+		}
+	}
+
+	public async validateCommit(apiContext: APIContext, commit: SingleCommit): Promise<boolean> {
+		// Validation step 4
+		const blockHeaderAtCommitHeight = await this._chain.dataAccess.getBlockHeaderByHeight(
+			commit.height,
+		);
+		if (!blockHeaderAtCommitHeight.id.equals(commit.blockID)) {
+			return false;
+		}
+
+		// Validation step 1
+		const doesCommitExistsInNonGossipedCommits = !!this._nonGossipedCommits
+			.get(commit.height)
+			?.some(
+				nonGossipedCommit =>
+					nonGossipedCommit.blockID === commit.blockID &&
+					nonGossipedCommit.validatorAddress === commit.validatorAddress,
+			);
+
+		const doesCommitExistsInGossipedCommits = !!this._gossipedCommits
+			.get(commit.height)
+			?.some(
+				gossipedCommit =>
+					gossipedCommit.blockID === commit.blockID &&
+					gossipedCommit.validatorAddress === commit.validatorAddress,
+			);
+
+		const doesCommitExist =
+			doesCommitExistsInGossipedCommits || doesCommitExistsInNonGossipedCommits;
+
+		if (doesCommitExist) {
+			return false;
+		}
+
+		// Validation Step 2
+		const maxRemovalHeight = await this._getMaxRemovalHeight();
+		if (commit.height <= maxRemovalHeight) {
+			return false;
+		}
+
+		// Validation Step 3
+		const { maxHeightPrecommitted } = await this._bftAPI.getBFTHeights(apiContext);
+		const isCommitInRange =
+			commit.height >= maxHeightPrecommitted - COMMIT_RANGE_STORED &&
+			commit.height < maxHeightPrecommitted;
```

Comment:
shouldn't this be less than equal?
```suggestion
			commit.height <= maxHeightPrecommitted;
```

Reply:
It's hard to understand end range in the LIP is included or not, are you sure about this?

Reply:
to my understanding, 
```ts
{bftModule.getBFTHeights().maxHeightPrecommitted - COMMIT_RANGE_STORED, ..., bftModule.getBFTHeights().maxHeightPrecommitted}
```
curly brackets `{ }` signify the set of values and the LIP says `Discard m if m.height is not in` the above range so I think its included. But we can clarify with @janhack

Reply:
Ishan's suggestion is correct. The notation {2, ..., 5} means the values 2, 3, 4, 5 so including the bounds.

Reply:
I noticed that the formatting of the relevant section in LIP 61 was broken and fixed it in the PR https://github.com/LiskHQ/lips/pull/125.

Reply:
done